### ISSUE TITLE
Implement cslib_string's extract_token(s)

### ISSUE THREAD

**Issue by CareyJWilliams** [ChoicescriptIDE/cslib#10]
Implementation of two routines in cslib_string that allow for splitting up a string based on a list of tokens/delimiters.

**extract_token**
Takes a reference to a global holding a tokenized string, along with a list of delimiters.
cslib_ret will equal the first token or "" if none (or the end of the string) is found.
The token should also be removed from the original ref var so that subsequent calls to extract_token can
work the tokens iteratively.
```
*create tokenised_string "token1:token2|token3:token4"
*temp tokens "|:"
*gosub_scene cslib_token "tokenized_string" tokens
${cslib_ret} == "token1"
And tokenized_string should be edited in place: "token2|token3:token4" 
```

**extract_tokens**
Would work by calling extract_token repeatedly.
We can check the array is large enough with arr_count, and should error with *bug in these cases? 
```
*create my_arr_count
*create my_arr_1
*create my_arr_2
*create my_arr_3
*gosub_scene cslib_string extract_tokens "VALUE1|VALUE2|VALUE3" "|" "my_arr"
${my_arr_1} == "VALUE1"
${my_arr_2} == "VALUE2"
${my_arr_3} == "VALUE3"
```

**Comment by CareyJWilliams** [comment]
I had written these already, but I seem to have lost them.
I could only recover the tests for extract_token:

```
*comment :::::: EXTRACT TOKEN tests ::::::
*gosub test_start "EXTRACT_TOKEN #1"
*temp result_expected "VAL=TRUE"
*set str_buffer "VAL=TRUE|VAL=FALSE|VAL=TRUE"
*gosub_scene cslib_string extract_token "str_buffer" "|"
*gosub test_finish result_expected cslib_ret

*gosub test_start "EXTRACT_TOKEN #2"
*temp result_expected "VAL=FALSE|VAL=TRUE"
*set str_buffer "VAL=TRUE|VAL=FALSE|VAL=TRUE"
*gosub_scene cslib_string extract_token "str_buffer" "|"
*gosub test_finish result_expected str_buffer

*gosub test_start "EXTRACT_TOKEN #3"
*temp result_expected "VAL=TRUE, VAL=FALSE, VAL=TRUE"
*temp result_actual ""
*set str_buffer "VAL=TRUE|VAL=FALSE|VAL=TRUE"
*label test_extact_token_3_loop
*gosub_scene cslib_string extract_token "str_buffer" "|"
*if cslib_ret != ""
	*set result_actual &("@{(result_actual = \"\") |, }${cslib_ret}")
	*goto test_extact_token_3_loop
*gosub test_finish result_expected result_actual
```

**Comment by diegoboston** [comment]
I'm not sure I understand why test EXTRACT_TOKEN no. 1 is different from no. 2.

I would prefer if we didn't modify the original string, as it will be a VERY slow character-by-character operation. I think extract_tokens (into an array) will be more efficient - we scan the source only once for all items. Yes, we'll have to check the count.

Shouldn't the separator tokens be only one char? Would it make things easier?

**Comment by CareyJWilliams** [comment]
> I'm not sure I understand why test EXTRACT_TOKEN no. 1 is different from no. 2.
> 
> I would prefer if we didn't modify the original string, as it will be a VERY slow character-by-character operation. I think extract_tokens (into an array) will be more efficient - we scan the source only once for all items. Yes, we'll have to check the count.
> 
> Shouldn't the separator tokens be only one char? Would it make things easier?

One is checking the extracted token in cslib_ret, the other is checking the string buffer was correctly manipulated.

No slower than a lot of the stuff already going on in cslib string. You're not always going to have a global array available to store results, or it might not be big enough. I think we need both options, but I'm open to other ideas on how to implement them.

They are, it's a token list, so it'll stop on any of the characters in that string. See str_tok in the C string library for the inspiration here.

**Comment by danspinola** [comment]
I think I get what Carey wants to do here.

I think splitting on a single separator is neater. 😬

I agree we should try to be non-destructive. Instead of returning the token, we could return the end index of the substring and have an optional parameter for offsetting the starting index. This way you know the first index and the last and you can call `cslib_string.substring` to extract an element. It would also allow using local variables by passing the value instead of a reference to a global var.

**split**
```
routine split(s: string, separators: string, start_pos: int =1) -> int:
	for each index, char in enumerated string:
		for each sep in separators:
			if (char == sep):
				return index - 1; // this will return the first match, not the closest one
	else, return 0;
```

**extract_tokens**
```
routine extract_tokens(s: string, separators: string, array_ref: string) -> void:
	int end_last_token = 0;
	int start_new_token = 1;
	int iterator = 1;
	
	while true:
		end_last_token = split(s, separators, start_new_token);
		if (end_last_token == 0):
			break;
		array_ref[iterator] = cslib_string.substring(s, start_new_token, end_last_token);
		interator++;
		start_new_token = end_last_token + 1;
```

**Comment by CareyJWilliams** [comment]
> I think splitting on a single separator is neater. 😬

It does split on single tokens (characters). If you mean that you don't see the point in allowing multiple separators, then I'll concede to that. I definitely do think it gives it more flexibility, but it's not a hill I'm willing to die on.

I think with your approach @danspinola, there is no point in the singular routine anymore? We may as well just call `cslib_string index` (or alias to it) -- not that that's a bad idea.

**Comment by danspinola** [comment]
I just realised I didn't use the `start_pos` in pseudocode of the split routine 😅. But you guys get the idea, I hope. 😁

Sure, I can see it now. `cslib_string index` would do it.

**Comment by CareyJWilliams** [comment]
Hey, I'm all for not writing code we don't need :)
Of course, index only works if we agree to limit the separator list to one character (which I'm happy to do).

We'd still benefit from extract_tokens (to array) though!

**Comment by CareyJWilliams** [comment]
Resolved by fc848444e5826b361177942a63f1abb57eae7fa4.
